src/email-providers/class-sendgrid.php

```diff
@@ -0,0 +1,464 @@
+	 * @return mixed
+	 */
+	public function get_campaign_id_from_create_result( array|false $result ): mixed {
+		if ( empty( $result ) ) {
+			return false;
+		}
+		$response = $result['response'];
+		if ( ! empty( $response && is_string( $response ) ) ) {
+			$body = (object) json_decode( $response );
+			return $body->id;
+		}
+		return false;
+	}
+
+
+	/**
+	 * Add subscriber to list
+	 *
+	 * @param string                       $list_id The list id.
+	 * @param string                       $email The email address.
+	 * @param array<array<string, string>> $custom_fields The custom fields.
+	 * @return array{
+	 *   response: mixed,
+	 *   http_status_code: int,
+	 * }|false  The response from the API.
+	 */
+	public function add_subscriber( string $list_id, string $email, array $custom_fields = [] ): array|false {
+		$sendgrid = $this->get_client();
+		if ( empty( $sendgrid ) ) {
+			return false;
+		}
+		$request_body           = (object) [];
+		$request_body->list_ids = [ $list_id ];
+		$user_object            = (object) [];
+		$user_object->email     = $email;
+		$request_body->contacts = [ $user_object ];
+		$response               = $sendgrid->client->marketing()->contacts()->put( $request_body );
+
+		return [
+			'response'         => $response->body(),
+			'http_status_code' => $response->statusCode(),
+		];
+	}
+
+	/**
+	 * Remove subscriber from list.
+	 *
+	 * @param string $list_id The list id.
+	 * @param string $email The email address.
+	 * @return array{
+	 *   response: mixed,
+	 *   http_status_code: int,
+	 * }|false  The response from the API.
+	 */
+	public function remove_subscriber( string $list_id, string $email ): array|false {
+		// TODO.
+		return [
+			'response'         => (object) [],
+			'http_status_code' => 500,
+		];
+	}
+
+	/**
+	 * Whether the provider manages from names.
+	 *
+	 * @return boolean
+	 */
+	public function provider_manages_from_names(): bool {
+		return true;
+	}
+
+	/**
+	 * Get the senders from the Sendgrid API.
+	 *
+	 * @return array<string>
+	 */
+	private function get_senders(): array {
+		$sendgrid = $this->get_client();
+		if ( empty( $sendgrid ) ) {
+			return [];
+		}
+		$response = $sendgrid->client->marketing()->senders()->get();
+
+		$body = json_decode( $response->body() );
+
+		$senders = [];
+		if ( ! is_array( $body ) ) {
+			return [];
+		}
+		foreach ( $body as $sender ) {
+			$senders[] = sprintf( '%s <%s>', $sender->from->name, $sender->from->email );
+		}
+		return $senders;
+	}
+
+	/**
+	 * Filters the from names to use Senders from the API.
+	 *
+	 * @param array<string> $from_names The existing from names.
+	 * @return array<string>
+	 */
+	public function filter_from_names( array $from_names ): array {
+		return $this->get_senders();
+	}
+
+	/**
+	 * Gets the html content for the newsletter.
+	 *
+	 * @param int $post_id The post id.
+	 * @return string|false
+	 */
+	private function get_content( $post_id ) {
```

Comment:
There are places where we don't use return types... is this intentional?

Reply:
not intentional